utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,602 @@
+
+instance ToJSON TextureDescriptorType where
+  toJSON     = String . textureDescriptorTypeToText
+  toEncoding = toEncoding . textureDescriptorTypeToText
+
+convertTextureDescriptorType :: TextureDescriptorType -> DescriptorType
+convertTextureDescriptorType = \case
+  Sampler2D -> DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER
+
+makeTextureDescriptorSetLayoutBinding
+  :: ShaderStage -> Texture -> (Int, DescriptorSetLayoutBinding)
+makeTextureDescriptorSetLayoutBinding stage Texture {..} =
+  ( set
+  , zero { binding         = fromIntegral binding
+         , descriptorType  = convertTextureDescriptorType type'
+         , descriptorCount = maybe 1 (V.sum . (fromIntegral <$>)) array
+         , stageFlags      = shaderStageFlagBits stage
+         }
+  )
+
+makeDescriptorSetLayoutBindings
+  :: ShaderStage
+  -> Vector Ubo
+  -> Vector Texture
+  -> Vector (Int, DescriptorSetLayoutBinding)
+makeDescriptorSetLayoutBindings stage ubos textures =
+  uboBindings <> textureBindings
+ where
+  uboBindings = V.map (makeUboDescriptorSetLayoutBinding stage) ubos
+  textureBindings =
+    V.map (makeTextureDescriptorSetLayoutBinding stage) textures
+
+makeDescriptorSetLayoutCreateInfos
+  :: Vector (Int, DescriptorSetLayoutBinding)
+  -> V.Vector (DescriptorSetLayoutCreateInfo '[])
+makeDescriptorSetLayoutCreateInfos bindings =
+  V.map makeDescriptorSetLayoutCreateInfo
+    . V.fromList
+    . M.elems
+    . M.unionWith (V.++) sets
+    $ setsMap
+ where
+  setSize = V.maximum . (fst <$>) $ bindings :: Int
+  sets :: Map Int (Vector DescriptorSetLayoutBinding)
+  sets = M.fromList . map (, []) $ [0 .. setSize]
+  setsMap :: Map Int (Vector DescriptorSetLayoutBinding)
+  setsMap =
+    M.fromList
+      . map (liftA2 (,) (fst . head) (V.fromList . (snd <$>)))
+      . groupBy ((==) `on` fst)
+      . sortOn fst
+      . V.toList
+      $ bindings
+
+makeDescriptorSetLayoutCreateInfo
+  :: Vector DescriptorSetLayoutBinding -> DescriptorSetLayoutCreateInfo '[]
+makeDescriptorSetLayoutCreateInfo bindings = zero { bindings = bindings }
+
+data VertexAttributeType
+  = Vec2
+  | Vec3
+  | Vec4
+  deriving (Show)
+
+vertexAttributeTypeFromText :: Text -> Maybe VertexAttributeType
+vertexAttributeTypeFromText = \case
+  "vec2" -> Just Vec2
+  "vec3" -> Just Vec3
+  "vec4" -> Just Vec4
+  _      -> Nothing
+
+vertexAttributeTypeToText :: VertexAttributeType -> Text
+vertexAttributeTypeToText = \case
+  Vec2 -> "vec2"
+  Vec3 -> "vec3"
+  Vec4 -> "vec4"
+
+instance FromJSON VertexAttributeType where
+  parseJSON = withTextMaybe "type" vertexAttributeTypeFromText
+
+instance ToJSON VertexAttributeType where
+  toJSON     = String . vertexAttributeTypeToText
+  toEncoding = toEncoding . vertexAttributeTypeToText
+
+convertVertexAttributeType :: VertexAttributeType -> (Word32, Format)
+convertVertexAttributeType = \case
+  Vec2 ->
+    (fromIntegral $ 2 * sizeOf (undefined :: Word32), FORMAT_R32G32_SFLOAT)
```

Comment:
Well, technically, vec's are `Float`. But you don't need `sizeOf` for that, since you specify 32 bits yourself in there.

Reply:
For documentation reasons I'd define something like `floatSize = 4` under a where-block and multiply by that.